libraries/eosiolib/core/eosio/crypto_ext.hpp

```diff
@@ -48,14 +49,183 @@ namespace eosio {
+      * Copy a serialized buffer to current point
+      *
+      * @param source - The source to be copied from
+      */
+      void copy_from(const std::vector<char>& source) {
+         eosio::check( source.size() == 2 * size, "size of souce buffer must be size of x + y" );
+         std::memcpy(x, source.data(), size);
+         std::memcpy(y, source.data() + size, size);
+      }
+   };
+
+   static constexpr size_t g1_coordinate_size = 32;
+   static constexpr size_t g2_coordinate_size = 64;
+   static constexpr size_t blake2f_result_size = 64;
+
+   /**
+    * Abstracts G1 point. Ensures sizes of x and y are 32
+    *
+    *  @ingroup crypto
+    */
+   struct g1_view : public point_view {
+      /**
+       * Construct a G1 point given x and y
+       *
+       * @param x_ - The x coordinate, a vector of chars
+       * @param y_ - The y coordinate, a vector of chars
+       */
+      g1_view(std::vector<char>& x_, std::vector<char>& y_)
+      : point_view(x_, y_)
+      {
+         eosio::check( size == g1_coordinate_size, "G1 coordinate size must be 32" );
+      }
+
+      /**
+       * Construct a point given a serialized point
+       *
+       * @param p - The serialized point
+       */
+      g1_view(std::vector<char>& p)
+      : point_view(p)
+      {
+         eosio::check( p.size() == g1_coordinate_size * 2, "G1 serialized size must be 64" );
+      };
+   };
+
+   /**
+    * Abstracts G2 point. Ensures sizes of x and y are 64
+    *
+    *  @ingroup crypto
+    */
+   struct g2_view : public point_view {
+      /**
+       * Construct a G2 point given x and y
+       *
+       * @param x_ - The x coordinate, a vector of chars
+       * @param y_ - The y coordinate, a vector of chars
+       */
+      g2_view(std::vector<char>& x_, std::vector<char>& y_)
+      : point_view(x_, y_)
+      {
+         eosio::check( size == g2_coordinate_size, "G2 coordinate size must be 64" );
+      }
+
+      /**
+       * Construct a point given a serialized point
+       *
+       * @param p - The serialized point
+       */
+      g2_view(std::vector<char>& p)
+      : point_view(p)
+      {
+         eosio::check( p.size() == g2_coordinate_size * 2, "G2 serialized size must be 128" );
+      };
+   };
+
+   /**
+    * Abstracts big integer.
+    *
+    *  @ingroup crypto
+    */
+   struct bigint {
+      /**
+       * Construct a bigint given a vector of chars (bytes)
+       *
+       * @param s - The source bytes
+       */
+      bigint(std::vector<char>& s)
+      :data(s.data()), size(s.size())
+      {
+      };
+
+      char* data;
+      uint32_t size;
+   };
+
+   /**
+    *  Addition operation on the elliptic curve `alt_bn128`
+    *
+    *  @ingroup crypto
+    *  @param op1 - operand 1
+    *  @param op2 - operand 2
+    *  @param result - result of the addition operation
+    *  @return -1 if there is an error otherwise 0
+    */
+   inline int32_t alt_bn128_add( const g1_view& op1, const g1_view& op2, g1_view& result) {
```

Comment:
For these we can `check` the internal return value and assert. And return a "result" from this instead of making it an in/out param.

Reply:
Will change.